### DISCUSSION TITLE
IntracellularRecordingsTable in Matlab

### DISCUSSION THREAD

**Discussion by dervinism** [NeurodataWithoutBorders/helpdesk#45] [Q&A]
Hi All,
I have a question about [IntracellularRecordingsTable](https://neurodatawithoutborders.github.io/matnwb/doc/+types/+core/IntracellularRecordingsTable.html). It seems that I cannot convert it to a plain table easily in Matlab. loadAll and getRow functions don't seem to work. Does it work differently in Python? I was expecting a similar functionality to ElectrodeTable of extracellular electrophysiology.

**Comment by lawrence-mbf** [comment]
Hi, 

There actually are overloaded functions for `getRow`, and `toTable` but they are under [DynamicTable](https://neurodatawithoutborders.github.io/matnwb/doc/+types/+hdmf_common/DynamicTable.html) though they should be useable by the IntracellularRecordingsTable.

I'm not sure how comparable the API for these methods are with PyNWB. I would assume no. That said, the documentation for each are almost always found under the `types.util.dynamictable` module (i.e., `getRow` is just a method forwarder for `types.util.dynamictable.getRow`) so hopefully the documentation there helps a bit more.

**Reply by lawrence-mbf** [reply]
A side-note, there is a `toTable` function call there. I need to regenerate the docs.

**Reply by dervinism** [reply]
If I run the [MatNWB Intracellular Electrophysiology tutorial](https://neurodatawithoutborders.github.io/matnwb/tutorials/html/icephys.html) up to the point where Hierarchical organization of recordings is described and then call
```
>> nwbfile.general_intracellular_ephys_intracellular_recordings.toTable()
```
I get an error:
```
Error using indexing
The specified key is not present in this container.

Error in types.untyped.Set/get (line 165)
                o{i} = obj.map(name{i});

Error in types.util.dynamictable.getRow>select (line 85)
    Vector = DynamicTable.vectordata.get(column);

Error in types.util.dynamictable.getRow (line 44)
    row{i} = select(DynamicTable, indexNames, ind);

Error in types.hdmf_common.DynamicTable/getRow (line 119)
        row = types.util.dynamictable.getRow(obj, id, varargin{:});

Error in types.util.dynamictable.nwbToTable (line 81)
matlabTable = [matlabTable DynamicTable.getRow( ...

Error in types.hdmf_common.DynamicTable/toTable (line 123)
        table = types.util.dynamictable.nwbToTable(obj, varargin{:});
```

**Reply by lawrence-mbf** [reply]
Ah, I see. This is an [AlignedDynamicTable](https://neurodatawithoutborders.github.io/matnwb/doc/+types/+hdmf_common/AlignedDynamicTable.html) so one would have to approach recreating these interfaces differently when they're injected. Unfortunately that will just have to be a limitation until I figure that out.

**Reply by dervinism** [reply]
Ok, I get it now. It would be much more convenient to be able to access the AlignedDynamicTable the same way as a regular DynamicTable. Afterall, it is aligned (or should be).

A related question. Hierarchical organisation of recordings does not necessesarily need to be based on simultaneous recordings. Regular recordings can be grouped in the same way in principle. Adding the name property would allow renaming the SimultaneousRecordingsTable. Do you intend to make hierarchical grouping more flexible?

**Reply by bendichter** [reply]
@dervinism , @oruebel is the expert of the schema here, but I believe the answer is that if you would like to group recordings that are not simultaneous you should do so at one of the next table levels: `Sequential RecordingsTable` or `RepetitionsTable` and just have a single recording for each entry of the `SimultaneousRecordingsTable`. Would either of these work for you?

**Reply by oruebel** [reply]
@bendichter yes, that is correct, for other groupings you would add additional metadata columns on the appropriate level of the temporal hierarchy. 

Having convenience functions for AlignedDynamicTable to be able to access rows and convert to a MatLab table would be nice. HDMF/PyNWB implement a number of these kind of convenience functionality. @lawrence-mbf just ping me if you have questions about AlignedDynamicTable

**Reply by dervinism** [reply]
Thanks! That makes sense. It was very helpful. I will also check how it works in Python.

**Comment by LeviGaetanoJZt** [comment]
In this case, I agree with @BabaraBrenes7mc that in Python, the IntracellularRecordingsTable can be accessed using PyNWB, and there are methods to convert it to a pandas DataFrame or numpy array. However, I cannot comment on the functionality in Matlab, as I do not have access to the latest information on the topic.

**Comment by MelissiaKorineckM0u** [comment]
IntracellularRecordingsTable is a specialized table format for intracellular electrophysiology data and may not have the same functionality as ElectrodeTable for extracellular electrophysiology. In Python, you can try using the NWB format (Neurodata Without Borders) for storing and manipulating intracellular recordings data, which is supported by the PyNWB library. It provides a convenient way to store, read and write electrophysiology data and metadata, and also offers tools for data analysis and visualization. You may want to consider using this format and library in your work.

**Comment by bendichter** [comment]
and the era of ChatGPT spam begins... sorry for the noise, @dervinism
